src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -344,8 +425,20 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
 		 */
 		if (!gotResults && expectResults)
 		{
-			queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
-									   &currentAffectedTupleCount);
+			/* use try/catch to maintain txn participant list during fail-fast errors */
+			PG_TRY();
+			{
+				queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
+										   &currentAffectedTupleCount);
+			}
+			PG_CATCH();
+			{
+				/* this placement is already aborted; remove it from participant list */
+				participantList = list_delete_ptr(participantList, taskPlacement);
```

Comment:
list_delete_ptr is O(n). Harmless enough here, but in future you might want to look at list_delete_cell instead (or use ilists.h style lists).